### PR TITLE
fix(DiffHighlightService): Skip unchanged char after pure addition/removal

### PR DESCRIPTION
**Pull request by mstv**
## Proposed changes

`DiffHighlightService.MarkDifference`:
- search the end of the diff back to `beginOffset` including
- avoid accessing characters out of range, i.e. before `beginOffset`

### Before

image

### After

image

## Test methodology 

- manual

## Merge strategy

I agree that the maintainer squash merge this PR (if the commit message is clear).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

**Comment by gerhardol** [comment]
The problem with removing this marker is that in some situations you do not see where the removal was done.
But I have no good example right now

**Comment by mstv** [comment]
> The problem with removing this marker is that in some situations you do not see where the removal was done.

The algorithm is so simple that the position is always clear.
It starts the highlight at the first different char and ends it at the last difference.

**Comment by RussKie** [comment]
@mstv could you help me understanding the difference between _before_ and _after_ screenshots - they look the same to me.

**Comment by gerhardol** [comment]
> @mstv could you help me understanding the difference between _before_ and _after_ screenshots - they look the same to me.

The paranthesis on line 132 is marked in the first screenshot.

I do not want to block this pr and will not protest if it is merged without my approval, but I recall there were some scenario where the extra mark was helpful. Have not reproduced yet.

**Comment by RussKie** [comment]
> The paranthesis on line 132 is marked in the first screenshot.

Ahh, now I see it, thanks. Now that I see the difference, I admit the "before" diff does look incorrect and the "after" looks correct as far as highlighting is concerned.

> I recall there were some scenario where the extra mark was helpful. Have not reproduced yet.

Would it make sense to revisit the implementation, if you manage to repro that scenario?

**Review by RussKie — APPROVED** [review]
👍 based on visual diff. I'm not familiar with the actual code.

**Comment by gerhardol** [comment]
> Would it make sense to revisit the implementation, if you manage to repro that scenario?

Yes, I am not opposed in merging this do not remember the scenario where the marker was helping.

**Comment by gerhardol** [comment]
My argument to keep the unchanged char is to see that a line is changed, to differ from completely new lines.
Especially if there are a block of lines slightly changed, it is hard to see if a line is completely new or that something is removed.

But that is discussed in #11851. (My latest suggestion there is to dim the foreground text for unchanged part, that would improve browsing changes. The bold part is not strictly needed then.)

**Pull request merged**
